### PR TITLE
Close AgileVentures/MetPlus_tracker#202

### PR DESCRIPTION
**Pull request by sidhene**
Created a home page for JobSeeker that shows personal information, AgencyPerson relationship information, and job search results. Includes test coverage.

I'm holding off on the rebase until I can get all the change requests in, since multiple rebasing has been causing problems.

### CODE REVIEW

**Comment by joaopapereira** [comment]
If this is going to be the final PR, nevertheless can be change please change the name to Close AgileVentures/MetPlus_tracker#202 so it can be tracked in waffle

File: app/controllers/job_seekers_controller.rb [hunk 1]
```diff
@@ -40,6 +40,13 @@ def update
     end
   end
 
+  def home
+    @jobseeker = JobSeeker.find(params[:id])
```
Comment (joaopapereira) [inline app/controllers/job_seekers_controller.rb]: What happen if we cannot find the JobSeeker?

File: app/models/agency_relation.rb [hunk 2]
```diff
@@ -8,4 +8,60 @@ class AgencyRelation < ActiveRecord::Base
   def self.in_role_of role_key
     where(agency_role_id: AgencyRole.find_by_role(AgencyRole::ROLE[role_key]).id)
   end
+
+  # Helper methods for associating job seekers with agency people
+  # These business rules are enforced:
+  # 1) A job developer can have only one case manager
+  # 2) A job developer can have only one job developer ('primary' JD)
+
+  def self.case_manager_for_job_seeker(job_seeker)
```
Comment (joaopapereira) [inline app/models/agency_relation.rb]: As we discussed on the meeting I do not believe this is the correct place to put these functions, they should be in AgencyPeople, because we are trying to retrieve AgencyPeople objects

File: app/models/agency_relation.rb [hunk 3]
```diff
@@ -8,4 +8,60 @@ class AgencyRelation < ActiveRecord::Base
...
+  end
+
+  def self.assign_job_developer_to_job_seeker(job_seeker, job_developer)
+    assign_agency_person_to_job_seeker(job_seeker, job_developer, :JD)
+  end
+
+  def self.assign_agency_person_to_job_seeker(job_seeker, agency_person, role_key)
+    raise 'Not a job seeker'   if not job_seeker.is_a? JobSeeker
+    raise 'Not an agency person' if not agency_person.is_a? AgencyPerson
+    raise 'AgencyPerson does not have appropriate role' if
```
Comment (joaopapereira) [inline app/models/agency_relation.rb]: We should not have this logic here. Isn't there a method in the AgencyPerson to check if that Person as a specific role?

File: app/models/agency_relation.rb [hunk 4]
```diff
@@ -8,4 +8,60 @@ class AgencyRelation < ActiveRecord::Base
...
+            agency_person.agency_roles.empty? ||
+            !agency_person.agency_roles.
+                  include?(AgencyRole.find_by_role(AgencyRole::ROLE[role_key]))
+
+    if (ap_relation = job_seeker.agency_relations.in_role_of(role_key)[0])
```
Comment (joaopapereira) [inline app/models/agency_relation.rb]: Is this statement correct? assignation in the if?

**Comment by joaopapereira** [comment]
Did you do integration tests?

**Comment by joaopapereira** [comment]
Not 100% convinced specially in the relationship part of the PR.
Please review my comments above

File: app/views/job_seekers/home.html.haml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    = @jobseeker.first_name
+    %br
+    %b Last Name:
+    = @jobseeker.last_name
+    %br
+    %b Status:
+    = JobSeekerStatus.find(@jobseeker.job_seeker_status_id).value
+    %br
+    %b Last Login:
+    = @jobseeker.last_sign_in_at.strftime('%x')
```
Comment (patmbolger) [inline app/views/job_seekers/home.html.haml]: If the user accepts the link in the confirmation email, but then subsequently never logs in, I think this attribute (last_sign_in_at) will be nil - in which case this statement will raise an exception.

Please look into that and harden this statement if needed.

File: app/views/jobs/index.html.haml [hunk 6]
```diff
@@ -10,16 +10,16 @@
...
+- else
 	- @jobs.each do |job|
 		%div{class: "well well-lg col-sm-offset-2 col-sm-8", id: "sub-list-id"}
-			= render job 
-%div{class: "col-sm-offset-4 col-sm-6"} 
-	= will_paginate 
\ No newline at end of file
+			= render job
+%div{class: "col-sm-offset-4 col-sm-6"}
+	= will_paginate
```
Comment (patmbolger) [inline app/views/jobs/index.html.haml]: I don't see any pagination on this page.  Please fix that.

Also, since we'll have many job seekers and multiple pages, we'll need to sort the displayed list by last name.

File: spec/controllers/job_seekers_controller_spec.rb [hunk 7]
```diff
@@ -168,6 +168,27 @@
...
+    end
+    it "returns http success" do
+      expect(response).to have_http_status(:success)
+    end
+
+    it "returns jobs posted since last login" do
+      @newjob = FactoryGirl.create(:job)
+      @newjob.assign_attributes(created_at: Time.now)
+      @jobseeker.assign_attributes(last_sign_in_at: (Time.now - 1.week))
+      expect(Job.new_jobs(@jobseeker.last_sign_in_at)).to include(@newjob)
```
Comment (patmbolger) [inline spec/controllers/job_seekers_controller_spec.rb]: This is a good test.

I suggest adding a few more jobs that were created within the last week.  And, also a few jobs that were created more than a week ago.  Then, the 'expect' can match returned jobs against an array that contains *only* the jobs that were created within the last week, and excludes the others.

File: app/models/agency_relation.rb (hunk 4 above)
Comment (patmbolger) [inline app/models/agency_relation.rb]: I wrote that code and, yes, the assignment within the if statement is intentional.  (perfectly valid Ruby)

File: app/views/job_seekers/home.html.haml [hunk 8]
```diff
@@ -0,0 +1,44 @@
+%h3 Your Information
+%div{:class => 'container'}
+  %div{class: 'div_align_left'}
```
Comment (patmbolger) [inline app/views/job_seekers/home.html.haml]: OK, I need to confess that I introduced the 'div_align_left' and 'div_align_right' styles back when I was first trying to get a column of field names (on the left) to line up with a column of values (on the right).  Subsequently I realized that this is a pretty bad way to do this for a number of reasons, including the fact that if one of the field descriptions is long and thus wraps, the next (lower) field and value pair will not be aligned with each other.

I have been using tables for this kind of thing from some point in the past on (for example, see app/views/agencies/_agency.html.haml).

If you don't mind a little extra work it would be good if you could convert this to use tables.  It is a much better technique and reduces that we'll repeat my original sin with further copy-and-paste.

File: app/models/agency_relation.rb (hunk 2 above)
Comment (patmbolger) [inline app/models/agency_relation.rb]: I struggled with this decision for the reason you cite, but chose to put these methods here because the two "heavy lifting" methods noted below are very tied up with the AgencyRelation model, and I did not feel it was appropriate for AgencyPerson methods to know that much about the internals of the AgencyRelation model.

The two "heavy lifting" methods are:

```
self.assign_agency_person_to_job_seeker
self.find_agency_person_for_job_seeker
```
How about we move the other methods to AgencyPerson and leave those two in AgencyRelation?

Comment (joaopapereira) [inline app/models/agency_relation.rb]: When I see objects being passed to Class functions so many times, for me automatically smells.
I believe this Class should be only the relationship and not a logic placeholder

File: app/models/agency_relation.rb (hunk 4 above)
Comment (joaopapereira) [inline app/models/agency_relation.rb]: I understand the Ruby allow you to do this, as almost all the languages do, but:
irb
2.2.1 :001 > if(ab = 100)
2.2.1 :002?>   puts 'bam'
2.2.1 :003?>   else
2.2.1 :004 >     puts 'bammm'
2.2.1 :005?>   end
(irb):1: warning: found = in conditional, should be ==
bam
 => nil 

I think the code should be as clean as possible and having assignations in the if's make it more cryptic, and we the only win there is 1 line.

Comment (patmbolger) [inline app/models/agency_relation.rb]: I agree it is a bit more cryptic.  I have no problem with having a separate assignment statement.

File: app/controllers/job_seekers_controller.rb (hunk 1 above)
Comment (sidhene) [inline app/controllers/job_seekers_controller.rb]: This is the JobSeeker's own home page. Although we don't have permissions yet (right?) it shouldn't be visible to anyone else. And they should get our standard 'you are not authorized' error with a redirect to the main page.

File: app/views/job_seekers/home.html.haml (hunk 5 above)
Comment (sidhene) [inline app/views/job_seekers/home.html.haml]: I've inserted an 'unless nil' clause

File: app/views/jobs/index.html.haml (hunk 6 above)
Comment (sidhene) [inline app/views/jobs/index.html.haml]: I need to revert this to whatever it is in our current branch. This file was not supposed to be changed

File: app/models/agency_relation.rb (hunk 2 above)
Comment (sidhene) [inline app/models/agency_relation.rb]: This function (and associated tests) have been moved

File: app/models/agency_relation.rb (hunk 3 above)
Comment (sidhene) [inline app/models/agency_relation.rb]: moved

File: app/models/agency_relation.rb (hunk 4 above)
Comment (sidhene) [inline app/models/agency_relation.rb]: Done yesterday in parking lot

File: app/controllers/job_seekers_controller.rb (hunk 1 above)
Comment (joaopapereira) [inline app/controllers/job_seekers_controller.rb]: I believe this is not a permission issue, if I go to /job_seekers/-1 what happens? 
or /job_seekers/10000000000?

File: spec/controllers/job_seekers_controller_spec.rb (hunk 7 above)
Comment (sidhene) [inline spec/controllers/job_seekers_controller_spec.rb]: I've only added one job (to test the negative case) but it failed properly (when I tried to use "assign_attributes" to change the created date, turns out you have to use "update_attributes"). I think this is good enough for now.

File: app/controllers/job_seekers_controller.rb (hunk 1 above)
Comment (amba178) [inline app/controllers/job_seekers_controller.rb]: Probably it will gives us NoRecordFound error, we could hand exceptions at application level and maybe redirect gracefully to any page we want. This can happen to other controllers so it would be nice to handle it at application level.

**Comment by sidhene** [comment]
@joaopapereira - Regarding the cucumber tests, I'm reviewing what we have now and they're all workflows, with several actions and usually a series of views. This story has just one view and not a lot of action, so I'm not clear on what your'e looking for other than confirming that the page loads (which we're already doing with Rspec) and the expected information is there. Will you be at standup today?

File: app/controllers/job_seekers_controller.rb (hunk 1 above)
Comment (joaopapereira) [inline app/controllers/job_seekers_controller.rb]: Maybe we can do that at application level, but nevertheless some sort of information should come out of here, like a string with the error that should be shown or something similar, that is the least in order to be able to have this setup at application level

**Comment by joaopapereira** [comment]
So some ideas come to mind:
* Now we have a link to access the edit form for job seeker, we should update the cukes that edit JobSeeker to come through this new page
* Are we testing presence and not presence of Job Developer and Case Manager display?
* Are we checking when there are no jobs no jobs popup and when there are jobs they show up?

File: app/views/job_seekers/home.html.haml (hunk 8 above)
Comment (sidhene) [inline app/views/job_seekers/home.html.haml]: I spent some time working on this and remembered that the reason why I used %div is because that's the only way I could get that to come out looking ok. If you really want this, we can try to solve it Monday, but I'm going to change it back.

**Comment by sidhene** [comment]
@joaopapereira - in the interest of getting this story and the next delivered, let's put these new cucumber tests in a different story.

**Comment by joaopapereira** [comment]
The story associated with this PR is:
As a registered job seeker
When I login to PETS
I should be directed to a personalized 'home page'

Is this story working?
Is this PR the latest code?

**Comment by sidhene** [comment]
Yes, except that we're using a navbar option instead of custom redirects on these stories.

File: app/models/agency.rb [hunk 9]
```diff
@@ -31,18 +31,24 @@ def agency_people_on_role role
...
     first.agency_people.pluck(:email)
   end
   ###################################################
 
-
   private
 
   def self.find_users_with_role(agency, role)
-    agency.agency_people_on_role role
+    users = []
```
Comment (joaopapereira) [inline app/models/agency.rb]: This change need to be reverted

File: spec/models/job_seeker_spec.rb [hunk 10]
```diff
@@ -21,22 +21,79 @@
...
-      expect(JobSeeker.acting_as? :user).to be true
-      expect(JobSeeker.acting_as? User).to  be true
-    end
-
-    it "returns false for anything other than supermodel" do
-      expect(JobSeeker.acting_as? :model).to be false
-      expect(JobSeeker.acting_as? String).to be false
-    end
-  end
+	context "#acting_as?" do
```
Comment (joaopapereira) [inline spec/models/job_seeker_spec.rb]: Is this change of indentation correct?

**Comment by joaopapereira** [comment]
@sidhene Please do the 2 changes I commented on.
After I will merge the PR

File: spec/models/job_seeker_spec.rb (hunk 10 above)
Comment (sidhene) [inline spec/models/job_seeker_spec.rb]: Yes, it was indented too far and I fixed it.

**Comment by sidhene** [comment]
Done

File: app/models/agency.rb (hunk 9 above)
Comment (sidhene) [inline app/models/agency.rb]: Done. Pushed.

File: app/models/agency.rb [hunk 11]
```diff
@@ -38,11 +38,18 @@ def self.all_agency_people_emails
   end
   ###################################################
 
-
   private
 
   def self.find_users_with_role(agency, role)
     agency.agency_people_on_role role
+    users = []
```
Comment (joaopapereira) [inline app/models/agency.rb]: Previously I say to revert this change, it was to remove what you added and keep what was there before.
You need to remove this added code, there is no need to you to even change this file

Comment (sidhene) [inline app/models/agency.rb]: @joaopapereira - that's not my code. It's either from the Agency work Patrick did mid-story, or the interim work that happened on other branches that I merged before placing the pull request. 

Are you asking me to remove the entire method? I tried commenting it out to see what would happen and it causes 30 test failures.

Comment (sidhene) [inline app/models/agency.rb]: Ok, it looks like the change is limited to inside this one method - a longer and shorter version. I swapped out to the shorter version and nothing's failing. I'm pretty sure this is Patrick's work, so I'm going to ask him what the difference is. It sounds like the shorter version is the one you want, so I'm going to check with him to ensure there's not a reason to use the longer one.

Comment (sidhene) [inline app/models/agency.rb]: Confirmed (shorter). Done. Pushed.

**Comment by joaopapereira** [comment]
Going to merge.
@sidhene do not forget to follow up with the 2 stories needed to end this story
